fix(`anvil`): set `storage.best_number` correctly

yash-atreya opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests.
-->

## Motivation

Closes #9053 + Closes https://github.com/foundry-rs/foundry/issues/8493

This arises when `--fork-url` and `--load-state` are used together.

<!--
Explain the context and why you're making that change. What is the problem
you're trying to solve? In some cases there is not a problem and this can be
thought of as being the motivation for your change.
-->

## Solution

- ~Introduce `account_at` in `ForkedStorage` which stores the state of accounts at specific block numbers.~
- ~Populate `account_at` using the provided state file.~ 

Solution is a lot simpler.

- We are correctly loading the data to the db: https://github.com/foundry-rs/foundry/blob/4012adefd376bd618d1348398c1da07224d2dace/crates/anvil/src/eth/backend/mem/mod.rs#L925
- However, the `storage.best_number` i.e head of the chain is not correctly set when `--load-state` is used with `--fork-url`. We continue to set `storage.best_number = state.block.number` which is incorrect as the `best_number` should be `fork_block_number`. 
- This issue is observed when `fork_block_number` > `state.block_number`. In this case while all state data has been correctly loaded to the db, we fetch via the provider due to this check. https://github.com/foundry-rs/foundry/blob/4012adefd376bd618d1348398c1da07224d2dace/crates/anvil/src/eth/api.rs#L683
- Setting `storage.best_number = fork_block_number` fixes this

- [x] Test

<!--
Summarize the solution and provide any necessary context needed to understand
the code change.
-->

grandizzy commented:
@yash-atreya I added a test here, basically dumping state from gh issue and reloading / make sure balance properly retrieved, feel free to include in PR if OK

https://github.com/foundry-rs/foundry/compare/yash/load-forked-state...grandizzy:test-for-9215

yash-atreya commented:
> @yash-atreya I added a test here, basically dumping state locally from gh issue recreate and reloading / make sure balance properly retrieved, feel free to include in PR if OK
> 
> [yash/load-forked-state...grandizzy:test-for-9215](https://github.com/foundry-rs/foundry/compare/yash/load-forked-state...grandizzy:test-for-9215)

@grandizzy thanks for this. 

I've added a more comprehensive test. ptal.

grandizzy commented:
@yash-atreya I suspect this is supposed to close https://github.com/foundry-rs/foundry/issues/8493 too? If so, could you please update description to include it? thank you

grandizzy commented on crates/anvil/src/eth/backend/mem/mod.rs:
```diff
@@ -907,19 +907,27 @@ impl Backend {
-            self.blockchain.storage.write().best_hash = best_hash;
+            if let Some((number, hash)) = fork_num_and_hash {
+                // If loading state file on a fork, set best number to the fork block number.
+                // Ref:
```
did you want to include ref to a gh issue here?

grandizzy reviewed (approved):
lgtm!
